Source/WebCore/platform/graphics/transforms/TransformState.cpp

```diff
@@ -232,6 +245,12 @@ void TransformState::flattenWithTransform(const TransformationMatrix& t, bool* w
             if (m_lastPlanarSecondaryQuad)
                 m_lastPlanarSecondaryQuad = t.mapQuad(*m_lastPlanarSecondaryQuad);
         }
+        if (m_tracking != DoNotTrackTransformMatrix) {
+            if (!m_trackedTransform)
```

Comment:
Have you considered reversing the logic? One less operation needed.

Reply:
```
diff --git a/Source/WebCore/platform/graphics/transforms/TransformState.cpp b/Source/WebCore/platform/graphics/transforms/TransformState.cpp
index 8bc1a7ec7f91..fefe5d4e0871 100644
--- a/Source/WebCore/platform/graphics/transforms/TransformState.cpp
+++ b/Source/WebCore/platform/graphics/transforms/TransformState.cpp
@@ -245,12 +245,6 @@ void TransformState::flattenWithTransform(const TransformationMatrix& t, bool* w
             if (m_lastPlanarSecondaryQuad)
                 m_lastPlanarSecondaryQuad = t.mapQuad(*m_lastPlanarSecondaryQuad);
         }
-        if (m_tracking != DoNotTrackTransformMatrix) {
-            if (!m_trackedTransform)
-                m_trackedTransform = makeUnique<TransformationMatrix>(t);
-            else
-                m_trackedTransform = makeUnique<TransformationMatrix>(t * *m_trackedTransform);
-        }
     } else {
         TransformationMatrix inverseTransform = valueOrDefault(t.inverse());
         if (m_mapPoint)
@@ -260,10 +254,16 @@ void TransformState::flattenWithTransform(const TransformationMatrix& t, bool* w
             if (m_lastPlanarSecondaryQuad)
                 m_lastPlanarSecondaryQuad = inverseTransform.projectQuad(*m_lastPlanarSecondaryQuad, wasClamped);
         }
-        if (m_tracking != DoNotTrackTransformMatrix) {
-            if (!m_trackedTransform)
-                m_trackedTransform = makeUnique<TransformationMatrix>(t);
-            else
+    }
+
+    if (m_tracking != DoNotTrackTransformMatrix) {
+        if (!m_trackedTransform)
+            m_trackedTransform = makeUnique<TransformationMatrix>(t);
+        else {
+            if (m_direction == ApplyTransformDirection) {
+                auto& trackedTransform = *m_trackedTransform;
+                trackedTransform = t * trackedTransform;
+            } else
                 m_trackedTransform->multiply(t);
         }
     }
```

That's what you had in mind?

Reply:
no, just:
if (not something) { foo } else { bar } -> if (something) { bar } else { foo }

Reply:
Combining the comments and further condensing reveals:

```
diff --git a/Source/WebCore/platform/graphics/transforms/TransformState.cpp b/Source/WebCore/platform/graphics/transforms/TransformState.cpp
index 8bc1a7ec7f91..d53722479ffd 100644
--- a/Source/WebCore/platform/graphics/transforms/TransformState.cpp
+++ b/Source/WebCore/platform/graphics/transforms/TransformState.cpp
@@ -245,12 +245,6 @@ void TransformState::flattenWithTransform(const TransformationMatrix& t, bool* w
             if (m_lastPlanarSecondaryQuad)
                 m_lastPlanarSecondaryQuad = t.mapQuad(*m_lastPlanarSecondaryQuad);
         }
-        if (m_tracking != DoNotTrackTransformMatrix) {
-            if (!m_trackedTransform)
-                m_trackedTransform = makeUnique<TransformationMatrix>(t);
-            else
-                m_trackedTransform = makeUnique<TransformationMatrix>(t * *m_trackedTransform);
-        }
     } else {
         TransformationMatrix inverseTransform = valueOrDefault(t.inverse());
         if (m_mapPoint)
@@ -260,14 +254,13 @@ void TransformState::flattenWithTransform(const TransformationMatrix& t, bool* w
             if (m_lastPlanarSecondaryQuad)
                 m_lastPlanarSecondaryQuad = inverseTransform.projectQuad(*m_lastPlanarSecondaryQuad, wasClamped);
         }
-        if (m_tracking != DoNotTrackTransformMatrix) {
-            if (!m_trackedTransform)
-                m_trackedTransform = makeUnique<TransformationMatrix>(t);
-            else
-                m_trackedTransform->multiply(t);
-        }
     }

+    if (m_trackedTransform)
+       *m_trackedTransform = (m_direction == ApplyTransformDirection) ? (t * *m_trackedTransform) : (*m_trackedTransform * t);
+    else if (m_tracking != DoNotTrackTransformMatrix)
+        m_trackedTransform = makeUnique<TransformationMatrix>(t);
+
     // We could throw away m_accumulatedTransform if we wanted to here, but that
     // would cause thrash when traversing hierarchies with alternating
     // preserve-3d and flat elements.
```

Rob likes this the most, so let's go for it :-)

Reply:
Fixed.